src/Common/Dispatcher/ListenerProvider.php

```diff
@@ -63,12 +69,14 @@ private function resolveListeners(string $role, array $config): void
             if (!$this->validateMacrosDefinition($definition)) {
                 continue;
             }
+            /** @psalm-var class-string $class */
             $class = $definition[self::DEFINITION_CLASS];
             $arguments = $definition[self::DEFINITION_ARGS] ?? [];
 
             $events = $this->findListenersInAttributes($class);
             try {
-                $listener = new $class(...$arguments);
+                $listener = (new Injector($this->container))->make($class, $arguments);
```

Comment:
The Injector instance should be creted once. Not in cycle.
You can create it in the constructor and store it instead of container instance.